src/coreclr/jit/emitxarch.cpp

```diff
@@ -423,6 +423,7 @@ bool emitter::AreFlagsSetToZeroCmp(regNumber reg, emitAttr opSize, genTreeOps tr
         case IF_RWR:
         case IF_RRD:
         case IF_RRW:
+        case IF_RWR_RRD_RRD:
```

Comment:
What about `RWR_RRD_ARD/SRD/MRD`? That is, the formats used for the `andn reg, reg, [mem]` encodings of the instruction?

Reply:
Can you explain how you derive the forms from the documentation? I'm not opposed to adding them I'd just like to understand the process a little better.

Reply:
the formats are basically just `IF_DST_OP1_OP2_OP...`

Where:
* `*WR` means "this is written"
* `*RW` means "this is read and written"
* `*RD" means "this is read"

For `*` you basically have:
* `R` which means it comes from a register
* `A`, `S`, `and `M` which indicates it comes from memory
  * `A` means "address mode"
  * `M` means "memory"
  * `S` means "stack"

https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/emitfmtsxarch.h goes into more detail and covers some of the edge cases, like `_CNS`

-------------------------

As from the architecture manual, these instructions are:
![image](https://user-images.githubusercontent.com/10487869/151429740-96ff3353-d9ea-4518-9dab-8dcc43c8a8d9.png)

And so we have `RWR_RRD_*RD`, where the last can be register (`R`) or any "from memory" concept (`A`, `M`, or `S`).

In codegen, you'll notice that we handle this "concept" centrally via `genHWIntrinsic_R_R_RM`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsiccodegenxarch.cpp#L1967-L1981

This flows down into some shared logic (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L1151-L1291) that looks if `op2` is contained (basically meaning it can come from memory in this case, or is otherwise specially handled by codegen) and how it is contained to decide if we want to use:
* `emitIns_SIMD_R_R_A` (which is the `RWR_RRD_ARD` format): https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L1225
* `emitIns_SIMD_R_R_C` (which is the `RWR_RRD_MRD` format): https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L1253
  * Yes the names don't exactly match up, but this is the existing "convention" 😄 
* `emitIns_SIMD_R_R_S` (which is the `RWR_RRD_SRD` format): https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L1270
* `emitIns_SIMD_R_R_R` (which is the `RWR_RRD_RRD` format): https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L1289

If you follow those down the `emitIns_SIMD_...` methods handle the RMW vs non-RMW difference, inserting a mov if necessary, https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/emitxarch.cpp#L6514-L6526
and then forward down to the `emitIns_...` method which creates the instruction descriptor and sets the format appropriately: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/emitxarch.cpp#L5137-L5162

For certain instructions, like `INS_addps`, there can be two possible paths and so two sets of valid formats:
* `addps xmm, xmm/m128` which is `RRW_*RD`, this is the `RMW` form used by hardware without AVX support
* `addps xmm, xmm, xmm/m128` which is `RWR_RRD_*RD`, this is the non-RMW` form used by hardware with AVX support

`BMI1/BMI2` instructions are generally only `non-RMW`, so you only have to worry about one "set" here

Reply:
Thank you. I've added all the memory formats. retested and updated the spmi results in the original post, no major changes.